File: Python/Observer pattern/subject_weather_data.py
from observers import *

class Subject:

    def registerObserver(self, observer):
        pass

    def removeObserver(self, observer):
        pass

    def notifyObservers(self):
        pass

class WeatherData(Subject):

    def __init__(self):
        self.temperature = 0
        self.humidity = 0
        self.pressure = 0
        self.observers = []

    def registerObserver(self, observer):
        self.observers.append(observer)
        
    def removeObserver(self, observer):
        for element in self.observers:
            if (id(element) == id(observer)):
                del element
                break

    def notifyObservers(self):
        for observer in self.observers:
            observer.update(self.temperature, self.humidity, self.pressure)
            
    def measurementsChanged(self):
        self.notifyObservers()

    def setMeasurements(self, temperature, humidity, pressure):
        self.temperature = temperature
        self.humidity = humidity
        self.pressure = pressure
        self.measurementsChanged()

    def getTemperatureStation(self):
        # get temperature from the station
        pass

    def getHumidityStation(self):
        # get humidity from the station
        pass

    def getPressureStation(self):
        # get pressure from the station
        pass
        
    #def returnObservers(self):
    #    return self.observers

    


    


        
        


    

File: ruby/observer_pattern/subject_weather_data.rb
require_relative 'observers'

class Subject

    def registerObserver(observer); end

    def removeObserver(observer); end

    def notifyObservers; end

end

class WeatherData < Subject

    def initialize
        @temperature = 0
        @humidity = 0
        @pressure = 0
        @@observers = []
    end

    def registerObserver(observer)
        @@observers.push(observer)
    end

    def removeObserver(observer)
        @@observers.each do |obs|
            if obs == observer
                @@observers.pop(observer)
            end
        end
    end

    def self.notifyObservers
        @@observers.each do |observer|
            observer.update(@temperature, @humidity, @pressure)
            return nil
        end
    end

    def self.measurementsChanged
        WeatherData.notifyObservers
    end

    def setMeasurements(temperature, humidity, pressure)
        @temperature = temperature
        @humidity = humidity
        @pressure = pressure
        WeatherData.measurementsChanged
    end

    def getTemperatureStation; end

    def getHumidityStation; end

    def getPressureStation; end

end